server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/CellIdSource.java

```diff
@@ -68,33 +84,71 @@ public SortedBinaryDocValues bytesValues(LeafReaderContext ctx) {
         throw new UnsupportedOperationException();
     }
 
-    /**
-     * The encoder to use to convert a geopoint's (lon, lat, precision) into
-     * a long-encoded bucket key for aggregating.
-     */
-    @FunctionalInterface
-    public interface GeoPointLongEncoder {
-        long encode(double lon, double lat, int precision);
+    /** Sorted numeric doc values for geo shapes */
+    protected static class GeoShapeCellValues extends AbstractSortingNumericDocValues {
+        private MultiGeoValues geoValues;
+        private int precision;
+        private GeoGridTiler tiler;
+
+        protected GeoShapeCellValues(MultiGeoValues geoValues, int precision, GeoGridTiler tiler) {
+            this.geoValues = geoValues;
+            this.precision = precision;
+            this.tiler = tiler;
+        }
+
+        protected void resizeCell(int newSize) {
+            resize(newSize);
+        }
+
+        protected void add(int idx, long value) {
+           values[idx] = value;
+        }
+
+        // for testing
+        protected long[] getValues() {
+            return values;
+        }
+
+        @Override
+        public boolean advanceExact(int docId) throws IOException {
+            if (geoValues.advanceExact(docId)) {
+                ValuesSourceType vs = geoValues.valuesSourceType();
+                MultiGeoValues.GeoValue target = geoValues.nextValue();
+                // TODO(talevy): determine reasonable circuit-breaker here
```

Comment:
do we plan to handle this before merging or can it wait?

Reply:
I consider this to be very harmful to a user since geo-data is often unwieldy and could easily result in someone requesting to tile a continent-sized shape using very high precision and causing dangerous memory pressure given how we are leveraging the `long[]` array to temporarily hold buckets.

I've opened #50720 to address this by accounting for the `long[]`'s bytes during bucket generation.

Reply:
oh and to answer the question: I hope to finish that PR before merging since it feels critical to me